src/water_cmd.cpp

```diff
@@ -1230,17 +1240,27 @@ static void DoDryUp(TileIndex tile)
  */
 void TileLoop_Water(TileIndex tile)
 {
-	if (IsTileType(tile, MP_WATER)) AmbientSoundEffect(tile);
+	if (IsTileType(tile, MP_WATER)) {
+		AmbientSoundEffect(tile);
+		if (IsNonFloodingWaterTile(tile)) return;
+	}
 
 	switch (GetFloodingBehaviour(tile)) {
-		case FLOOD_ACTIVE:
+		case FLOOD_ACTIVE: {
+			bool continue_flooding = false;
 			for (Direction dir = DIR_BEGIN; dir < DIR_END; dir++) {
 				TileIndex dest = AddTileIndexDiffCWrap(tile, TileIndexDiffCByDir(dir));
 				/* Contrary to drying up, flooding does not consider MP_VOID tiles. */
 				if (!IsValidTile(dest)) continue;
 				/* do not try to flood water tiles - increases performance a lot */
 				if (IsTileType(dest, MP_WATER)) continue;
 
+				/* Buoys and docks cannot be flooded, and when removed turn into flooding water. */
+				if (IsTileType(dest, MP_STATION) && (IsBuoy(dest) || IsDock(dest))) continue;
+
+				/* This neighbour tile might be floodable later if the tile is cleared, so allow flooding to continue. */
+				continue_flooding = true;
+
 				/* TREE_GROUND_SHORE is the sign of a previous flood. */
 				if (IsTileType(dest, MP_TREES) && GetTreeGround(dest) == TREE_GROUND_SHORE) continue;
```

Comment:
```suggestion
				/* TREE_GROUND_SHORE is the sign of a previous flood. */
				if (IsTileType(dest, MP_TREES) && GetTreeGround(dest) == TREE_GROUND_SHORE) continue;

				/* Industries on sea cannot be flooded, and when removed turn into flooding water. */
				if (IsTileType(dest, MP_INDUSTRY) && GetWaterClass(dest) == WATER_CLASS_SEA) continue;

				/* This neighbour tile might be floodable later if the tile is cleared, so allow flooding to continue. */
				continue_flooding = true;
```
`continue_flooding = true;` should be done after `TREE_GROUND_SHORE`. I also added industry tiles and the oilrig station.

Reply:
It shouldn't be done after TREE_GROUND_SHORE. I already did testing with that.